### ISSUE TITLE
LXRHash has a high cache hit rate with simple LRU cache

### ISSUE THREAD

**Issue by WhoSoup** [pegnet/LXR256#44]
tl;dr: if you want to prevent a high cpu cache hit rate with a very small amount of space, we need to re-order the input

I wanted to test the effect of CPU cache sizes on the performance of LXR Hash. The cache line of i7 processors is **64 bytes** big, meaning it caches that much sequential memory in one request. 

For the LRU cache, I used: https://github.com/hashicorp/golang-lru

My methodology:
1. I saved the **index** of every single request to ByteMap
2. I played through the indexes and got the specific cache line by `floor(index / 64)`
3. If that cache line was present in the cache, it counts as a hit
4. If that cache line is not present, it is added.
5. Divide the number of hits by the total requests to ByteMap

It turns out that when performing a mining run of 1000 sequential nonces appended to a base, the cache hit rate is about 50%, and this is true for a cache size of only **128 KB**, meaning that CPUs can potentially fit that into L2 cache opposed to L3, and you can speed up the algorithm as much as 50% with a very small amount of cache. (please note that the 50% hit rate holds true for 1,000 runs as it does for 100,000 runs)

Using random input, the cache hit rate is close to zero, growing a fraction the larger the cache is as you would expect.

![cache hit freq](https://i.imgur.com/BhLvvdZ.png)

So I went a step further and implemented a pseudo-random function that shuffles the input in a deterministic manner. This resulted in a similar cache hit rate as the random input while performing a mining run.

This behavior is backed up by empirical evidence:
```
BenchmarkHash/random_input-8         	   10000	    157108 ns/op	     288 B/op	       2 allocs/op
BenchmarkHash/run_input-8            	   20000	     98155 ns/op	     288 B/op	       2 allocs/op
```

There's a roughly 40% speed increase on sequential data opposed to random data. This can be accomplished using a very small amount of cache.

For our specific purposes, this can also be avoided by, yet again, running faststep and step backward (since the different nonce will have a cascading effect on the static opr hash). For a more generalized solution, the input can be pseudo-shuffled in an algorithm like:
```
	for i := 0; i < len(src); i++ {
		swp := i*int(src[i]) + (len(src)-1-i)*int(src[(len(src)-1-i)%len(src)])
		swp = swp % len(src)
		src[i], src[swp] = src[swp], src[i]
	}
```
(though ideally something that produces a bit better results)

**Comment by PaulSnow** [comment]
I am not quite following the methodology....  I had done some different tests, and didn't find any pattern in the indexes to the ByteMap that suggested an LRU cache would do much for us.

The suggested randomization works against the src data, but our src data is a hash, so it is already very close to random, so I am not sure what a randomization step will do for us.

**Comment by WhoSoup** [comment]
> I am not quite following the methodology.... I had done some different tests, and didn't find any pattern in the indexes to the ByteMap that suggested an LRU cache would do much for us.

I'm not looking at individual indexes but rather groups of 64 bytes of indexes, the way the cpu cache works. I'm not trying to optimize it, just replicating what the CPU does



> The suggested randomization works against the src data, but our src data is a hash, so it is already very close to random, so I am not sure what a randomization step will do for us.

it only works over successive runs, for which the first 32 bytes are always the same. the cache carries over for each hash

**Comment by PaulSnow** [comment]
Yeah, the source data stays in the cache. But for us that is 40 bytes more or less. 

Are you saying  that running the LXRHash is 40% faster if we don't reorder the inputs?

**Comment by PaulSnow** [comment]
It is actually step 3 that I don't quite get.  A hit on a cache line is not really a hit unless the address in the cache exactly matches the address you are trying to access...

**Comment by PaulSnow** [comment]
In effect, caches are Maps, and in the case of a cache that provides 64 bytes on a hit, then the byte accessed has to be a byte within those 64 bytes.  If it isn't, then you put a different 64 bytes in the cache, and toss the last recently used 64 bytes from the cache, and then continue.

**Comment by WhoSoup** [comment]
> Yeah, the source data stays in the cache. But for us that is 40 bytes more or less.
> 
> Are you saying that running the LXRHash is 40% faster if we don't reorder the inputs?

yes, since the input stays relatively the same, the parts of the bytemap that are accessed are also similar

**Comment by PaulSnow** [comment]
No, the fast pass randomizes the indexing

**Comment by WhoSoup** [comment]
> It is actually step 3 that I don't quite get. A hit on a cache line is not really a hit unless the address in the cache exactly matches the address you are trying to access...

nope, the cpu caches 64 bytes at a time when pulling a single address from memory. anything from a[0] to a[63] falls under the same ram request

**Comment by PaulSnow** [comment]
> nope, the cpu caches 64 bytes at a time when pulling a single address from memory. anything from a[0] to a[63] falls under the same ram request

That's true, but we nearly never get another ByteMap request within that 64 byte window, so caching 64 bytes per address buys the LXRHash nearly nothing.

**Comment by PaulSnow** [comment]
Let me write a test...

**Comment by WhoSoup** [comment]
> That's true, but we nearly never get another ByteMap request within that 64 byte window, so caching 64 bytes per address buys the LXRHash nearly nothing.

Yeah but a 128 KiB CPU cache has the last 2048 used windows. It just so happens that the LXR Hash re-uses ones inside that range. It does NOT work with a 64 KiB cache, so somewhere between every 1024 and 2048 requests are ones made to the same cache line

**Comment by WhoSoup** [comment]
> No, the fast pass randomizes the indexing

nope! the fast pass prevents someone from pre-calculating a part of the input, since the fast pass changes the initial counters based on the length of the input. Fastpass does not touch the input in any way (`v2` is in the parameter but it's not actually used)

**Comment by WhoSoup** [comment]
> > No, the fast pass randomizes the indexing
> 
> nope! the fast pass prevents someone from pre-calculating a part of the input, since the fast pass changes the initial counters based on the length of the input. Fastpass does not touch the input in any way (`v2` is in the parameter but it's not actually used)

out of curiosity, i changed fastpass to:
```
	faststep := func(v2 uint64, idx uint64) {
		as = idx<<1 ^ idx>>3 ^ as<<7 ^ as>>5
		s1 = s1<<9 ^ s1>>3 ^ as
		hs[idx] = s1 ^ as ^ v2
		as, s1, s2, s3 = s3, as, s1, s2
	}
```
and this will also do the trick of preventing caching. (i just added `^ v2` to the second to last line) (I'm getting 4-5% cache hit rate instead of 50%)

**Comment by PaulSnow** [comment]
That is exactly right, Who!  For some reason I didn't involve v2, but intended to do so. Sorry I was slow to understand today, but I had assumed v2 was used in faststep.

Personally, I think the first line might be better:
```
as = idx<<1 ^ idx>>3 ^ as<<7 ^ as>>5 ^ v2
```
Since as is a factor in s1 and hs[idx]

**Comment by PaulSnow** [comment]
So continued testing on this new issue raised by @WhoSoup , the limitations of xor/shift random number generators are quite evident.  For some ranges of nonce values, the right shifts and the right order of state modifications work pretty well to set the sequence in a low 4 MB cache setting with 25 bit ByteMaps.

The problem is that other nonce sequences, syncopation with the period of the xor/shift hash results in many, many cache hits.

The solution from the original paper on xor shift hashes is to introduce some other source of randomness.  So I suggest we do *one* ByteMap lookup per byte.  This produces uniform results over a wide set of nonce sequences that I have tested.

See #47 for the suggested code changes, and the updates to the unit tests.
